krm-functions/configinject-fn/fn/function.go

```diff
@@ -139,84 +140,118 @@ func (f *FnR) desiredOwnedResourceList(forObj *fn.KubeObject) (fn.KubeObjects, e
-			}
-			for _, ref := range dep.Spec.Injectors {
-				/*
-					for _, o := range rl.Items {
-						fn.Logf("configinject resource apiVersion: %s kind: %s\n", o.GetAPIVersion(), o.GetKind())
+			prName := fmt.Sprintf("%s-%s", pr.Spec.RepositoryName, pr.Spec.PackageName)
+
+			latestPR, ok := prmap[prName]
+			if !ok {
+				// we initialize hte prmap independent of the status
+				prmap[prName] = pr
+			} else {
+				// only update if the PR is published
+				if porchv1alpha1.LifecycleIsPublished(pr.Spec.Lifecycle) {
+					if !porchv1alpha1.LifecycleIsPublished(latestPR.Spec.Lifecycle) {
+						// if latest entry in the prmap is not publish but the new pr is published we update the map
+						prmap[prName] = pr
+					} else {
+						// both the latest pr and the new pr are published
+						// update the map with the latest pr
+						// if the revision of the new pr is better than the one of the latest pr in the map
+						if pr.Spec.Revision > latestPR.Spec.Revision {
+							prmap[prName] = pr
+						}
 					}
-				*/
-				fn.Logf("configinject dependency gvk: %s\n", gvk.String())
+				}
+			}
+		}
+	}
+
+	for _, pr := range prmap {
+		if !porchv1alpha1.LifecycleIsPublished(pr.Spec.Lifecycle) {
+			msg := fmt.Sprintf("configinject dependency not ready: the package %s in repo %s is not published\n", pr.Spec.PackageName, pr.Spec.RepositoryName)
+			fn.Logf("%s\n", msg)
+			// if 1 package is not ready we fail fast
+			return nil, fmt.Errorf("%s", msg)
+		}
+	}
+
+	// at this stage all packages are published
+	for _, pr := range prmap {
+		// get the package resources of the revision
+		prr := &porchv1alpha1.PackageRevisionResources{}
+		if err := f.Get(ctx, types.NamespacedName{Namespace: pr.Namespace, Name: pr.Name}, prr); err != nil {
+			return nil, err
+		}
+		// get the resource list from the package
+		rl, err := kptrl.GetResourceList(prr.Spec.Resources)
+		if err != nil {
+			return nil, err
+		}
+		// get the kptfile from the resourcelist to check condition status
+		kfko := rl.Items.GetRootKptfile()
+		if kfko == nil {
+			return nil, fmt.Errorf("mandatory Kptfile is missing from the package %s, repo %s", pr.Spec.PackageName, pr.Spec.RepositoryName)
+		}
+		kf := kptfilelibv1.KptFile{Kptfile: kfko}
 
-				depObjs := rl.Items.Where(fn.IsGroupVersionKind(ref.GroupVersionKind()))
-				if len(depObjs) == 0 {
-					fn.Logf("configinject dependency not ready: the package %s in repo %s does not contain a resource with %s\n", pr.Spec.PackageName, pr.Spec.RepositoryName, gvk.String())
-					return nil, fmt.Errorf("dependency not ready: the package %s in repo %s does not contain a resource with %s", pr.Spec.PackageName, pr.Spec.RepositoryName, gvk.String())
+		// get the dependency objects in the package and check its status
+		gvk := schema.GroupVersionKind{
+			Group:   nephiodeployv1alpha1.GroupVersion.Group,
+			Version: nephiodeployv1alpha1.GroupVersion.Version,
+			Kind:    nephiodeployv1alpha1.UPFDeploymentKind,
```

Comment:
Eventually this will need to be generic, but that's R1++ or R2 I guess?

Reply:
yes I need to think how to do this.